# Задача линейного программирования

**Выполнил:** Одилов Нодир  
**Вариант:** 2  
**Метод решения:** Python, `scipy.optimize.linprog`

### Исходные данные

$$
c = (-2, 1, 5, -3)
$$

$$
A =
\begin{pmatrix}
1 & -1 & 2 & 0 \\
0 & 3 & 1 & -2 \\
4 & 1 & 3 & 1
\end{pmatrix}
$$

$$
b = (8, 7, 15)
$$

Необходимо решить прямую задачу линейного программирования, построить двойственную задачу и проверить выполнение теоремы двойственности.

## 1. Прямая задача

Исходная задача имеет вид:

$$
F = -2x_1 + x_2 + 5x_3 - 3x_4 \rightarrow \max
$$

при ограничениях:

$$
\begin{cases}
x_1 - x_2 + 2x_3 \leq 8, \\
3x_2 + x_3 - 2x_4 \leq 7, \\
4x_1 + x_2 + 3x_3 + x_4 \leq 15, \\
x_1, x_2, x_3, x_4 \geq 0.
\end{cases}
$$

In [34]:
import numpy as np
from scipy.optimize import linprog

In [35]:
c = np.array([-2, 1, 5, -3], dtype=float)

A = np.array([
    [1, -1, 2,  0],
    [0,  3, 1, -2],
    [4,  1, 3,  1]
], dtype=float)

b = np.array([8, 7, 15], dtype=float)

print("c =", c)
print("A =")
print(A)
print("b =", b)

c = [-2.  1.  5. -3.]
A =
[[ 1. -1.  2.  0.]
 [ 0.  3.  1. -2.]
 [ 4.  1.  3.  1.]]
b = [ 8.  7. 15.]


In [36]:
# Решение прямой задачи

result_primal = linprog(
    c=-c,                  # linprog минимизирует, поэтому для max используем -c
    A_ub=A,
    b_ub=b,
    bounds=[(0, None)] * 4,
    method="highs"
)

print("Успешно решено:", result_primal.success)
print(
    "Оптимальное решение x =",
    [round(float(x), 6) for x in result_primal.x]
)
print(
    "Максимальное значение F =",
    round(float(-result_primal.fun), 6)
)

Успешно решено: True
Оптимальное решение x = [0.0, 0.857143, 4.428571, 0.0]
Максимальное значение F = 23.0


### Результат прямой задачи

Полученное оптимальное решение:

$$
x^*=
\left(
0,\frac{6}{7},\frac{31}{7},0
\right)
\approx
(0,\;0.857143,\;4.428571,\;0).
$$

Оптимальное значение целевой функции:

$$
F_{\max}=23.
$$

### Проверка допустимости решения

Подставим найденное решение

$$
x^*=
\left(
0,\frac67,\frac{31}{7},0
\right)
$$

в ограничения прямой задачи.

Первое ограничение:

$$
0-\frac67+2\cdot\frac{31}{7}
=
\frac{56}{7}
=
8 \leq 8.
$$

Второе ограничение:

$$
3\cdot\frac67+\frac{31}{7}
=
\frac{49}{7}
=
7 \leq 7.
$$

Третье ограничение:

$$
\frac67+3\cdot\frac{31}{7}
=
\frac{99}{7}
\approx14.142857\leq15.
$$

Условия неотрицательности также выполняются:

$$
x_1,x_2,x_3,x_4\geq0.
$$

Следовательно, найденное решение является допустимым.

## 2. Двойственная задача

Для прямой задачи вида

$$
\max \; c^T x
$$

при условиях

$$
Ax \leq b, \qquad x \geq 0
$$

двойственная задача имеет вид

$$
\min \; b^T y
$$

при условиях

$$
A^T y \geq c, \qquad y \geq 0.
$$

Матрица коэффициентов прямой задачи:

$$
A=
\begin{pmatrix}
1 & -1 & 2 & 0\\
0 & 3 & 1 & -2\\
4 & 1 & 3 & 1
\end{pmatrix}.
$$

Её транспонированная матрица:

$$
A^T=
\begin{pmatrix}
1 & 0 & 4\\
-1 & 3 & 1\\
2 & 1 & 3\\
0 & -2 & 1
\end{pmatrix}.
$$

Так как

$$
A^Ty\geq c,
$$

каждая строка матрицы $A^T$ образует одно ограничение двойственной задачи.

Для варианта 2 двойственная задача имеет вид:

$$
G = 8y_1 + 7y_2 + 15y_3 \rightarrow \min
$$

при ограничениях:

$$
\begin{cases}
y_1 + 4y_3 \geq -2, \\
-y_1 + 3y_2 + y_3 \geq 1, \\
2y_1 + y_2 + 3y_3 \geq 5, \\
-2y_2 + y_3 \geq -3, \\
y_1, y_2, y_3 \geq 0.
\end{cases}
$$

In [37]:
# Решение двойственной задачи

result_dual = linprog(
    c=b,
    A_ub=-A.T,            # A^T y >= c  ->  -A^T y <= -c
    b_ub=-c,
    bounds=[(0, None)] * 3,
    method="highs"
)

print("Успешно решено:", result_dual.success)
print(
    "Оптимальное решение y =",
    [round(float(y), 6) for y in result_dual.x]
)
print(
    "Минимальное значение G =",
    round(float(result_dual.fun), 6)
)

Успешно решено: True
Оптимальное решение y = [2.0, 1.0, 0.0]
Минимальное значение G = 23.0


### Результат двойственной задачи

Полученное оптимальное решение:

$$
y^*=(2,\;1,\;0).
$$

Оптимальное значение целевой функции:

$$
G_{\min}=23.
$$

## 3. Проверка теоремы сильной двойственности

Для прямой задачи было получено:

$$
x^* = (0,\;0.857143,\;4.428571,\;0)
$$

$$
F_{\max} = 23
$$

Для двойственной задачи:

$$
y^* = (2,\;1,\;0)
$$

$$
G_{\min} = 23
$$

Следовательно,

$$
F_{\max} = G_{\min} = 23.
$$

Таким образом, выполняется теорема двойственности линейного программирования.

## 4. Проверка условий дополняющей нежёсткости

Для оптимальных решений прямой и двойственной задач проверим условия дополняющей нежёсткости.

In [38]:
# Проверка условий дополняющей нежёсткости

x_opt = result_primal.x
y_opt = result_dual.x

primal_slack = b - A @ x_opt
dual_slack = A.T @ y_opt - c

print(
    "Запасы ограничений прямой задачи:",
    [round(float(v), 6) for v in primal_slack]
)

print(
    "Запасы ограничений двойственной задачи:",
    [round(float(v), 6) for v in dual_slack]
)

print(
    "y_i * primal_slack_i =",
    [round(float(v), 6) for v in y_opt * primal_slack]
)

print(
    "x_j * dual_slack_j =",
    [round(float(v), 6) for v in x_opt * dual_slack]
)

Запасы ограничений прямой задачи: [0.0, 0.0, 0.857143]
Запасы ограничений двойственной задачи: [4.0, 0.0, 0.0, 1.0]
y_i * primal_slack_i = [0.0, 0.0, 0.0]
x_j * dual_slack_j = [0.0, 0.0, 0.0, 0.0]


Все произведения в условиях дополняющей нежёсткости равны нулю.

Для прямой задачи первые два ограничения являются активными, а третье имеет запас

$$
\frac{6}{7}\approx0.857143.
$$

При этом соответствующая двойственная переменная равна

$$
y_3=0.
$$

Для двойственной задачи первое и четвёртое ограничения имеют положительные запасы, при этом

$$
x_1=x_4=0.
$$

Таким образом, условия дополняющей нежёсткости выполняются полностью, что дополнительно подтверждает оптимальность найденных решений.

## 5. Вывод

В работе была рассмотрена задача линейного программирования варианта 2.

Для прямой задачи было получено оптимальное решение

$$
x^*=
\left(
0,\frac67,\frac{31}{7},0
\right),
$$

при котором

$$
F_{\max}=23.
$$

Для прямой задачи была построена соответствующая двойственная задача. Её оптимальное решение:

$$
y^*=(2,1,0),
$$

при котором

$$
G_{\min}=23.
$$

Получено равенство

$$
F_{\max}=G_{\min}=23,
$$

что подтверждает выполнение теоремы сильной двойственности.

Дополнительно были проверены ограничения прямой задачи и условия дополняющей нежёсткости. Все проверки выполнены успешно.

Таким образом, найденные решения являются допустимыми и оптимальными.